### DISCUSSION TITLE
Bug in GenericAPIView where Browsable API fails to run if serialized nested dictionary

### DISCUSSION THREAD

**Discussion by dracarys-github** [encode/django-rest-framework#7959] [Potential Issue]
DRF Browsable API fails to run if the serialized data contains nested dictionaries when the both the get and post method is written:

Data being serialized is in the following format:
```
[{
   pk:1,
   item_name: "laptop"
   item_cat_name: {id:1,name:"techonology"}
}]
```

1) It works with API View
1) It works with GenericAPIView if only GET is written and not POST

This does not work:

```
class Item_Mast_ListView(GenericAPIView):
    queryset = Item_Mast.objects.all().select_related("item_cat_mast")
    serializer_class = Item_Mast_Serializer

    def get(self, request, *args, **kwargs):
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset,many=True)
        return Response(serializer.data)

    def post(self, request, *args, **kwargs):
        serializer = self.get_serializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        serializer.save()
        return Response(serializer.data, status=status.HTTP_201_CREATED)
```

This works
```
class Item_Mast_ListView(GenericAPIView):
    queryset = Item_Mast.objects.all().select_related("item_cat_mast")
    serializer_class = Item_Mast_Serializer

    def get(self, request, *args, **kwargs):
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset,many=True)
        return Response(serializer.data)
```

This also works

```
class Item_Mast_ListView(APIView):
    def get(self, request, *args, **kwargs):
        queryset = Item_Mast.objects.all().select_related("item_cat_mast")
        serializer = Item_Mast_Serializer(queryset,many=True)
        return Response(serializer.data)

    def post(self, request, *args, **kwargs):
        serializer = Item_Mast_Serializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        serializer.save()
        return Response(serializer.data, status=status.HTTP_201_CREATED)
```

For more detailed explanation, you can refer to this stackoverflow post: https://stackoverflow.com/questions/67329800/drf-browsable-api-make-sure-it-works-when-dictionary-is-passed-to-a-get-respon
For a reproducible example you can use this project: [https://github.com/dracarys-github/sample_SO_ref](url)
(Just run the project and go to the link localhost:8000/item_mast  to reproduce the error) 

**UPDATE**
It was found that the POST method adds extra inputs and a form when using GenericAPIView, I think DRF Web API should be made compatible to handle dictionaries being serialised.

**Comment by jerinpetergeorge** [comment]
from our [discussion](https://chat.stackoverflow.com/rooms/231937/discussion-between-jpg-and-dracarys)

> The root cause of the issue is, DRF expects the to_representation(..) method to return a single value like string or integer, in your case, it is a `dict`.
> 
> If you change the **`to_representation(...)`** to return a **_single value_**, there will not be any errors

Personally, I don't think this is an issue with the DRF or Browsable API, rather than an implementation issue.

**Reply by dracarys-github** [reply]
I think this is a potential issue solely for the reason that Django is able to handle Nested serializers. So I don't think it should have a problem of serialising dictionaries when the user returns that in the to_representation method.

As we discussed in our discussion on stackoverflow, this seems to be an issue just with GenericViewAPI, and viewsets etc and not APIView. This is because the GenericViewApi renders additional form fields for POST on the browsable API. 

So it isn't an issue with rendering dictionaries as much as something not handled when the additional form elements are used

**Reply by lovelydinosaur** [reply]
Could you be more clear in what you mean when you say "This does not work"?

For the browsable API, including an example with a screenshot might help explain the issue.

**Reply by jerinpetergeorge** [reply]
The minimal reproducible example is here,

```python
# models.py
class Musician(models.Model):
    name = models.CharField(max_length=50)

    def __str__(self):
        return f'{self.name}'


class Album(models.Model):
    artist = models.ForeignKey(Musician, on_delete=models.CASCADE)
    name = models.CharField(max_length=100)

    def __str__(self):
        return f'{self.name} : {self.artist}'

# serializers.py
class CustomField(serializers.RelatedField):
    def to_representation(self, value):
        return {"id": value.pk, "name": value.name}


class AlbumSerializer(serializers.ModelSerializer):
    artist = CustomField(queryset=Musician.objects.all())

    class Meta:
        fields = '__all__'
        model = Album

# views.py
class AlbumViewSet(viewsets.ModelViewSet):
    serializer_class = AlbumSerializer
    queryset = Album.objects.all()
```

and the complete error traceback can be found [here](https://dpaste.com/C3RCR83XR)

**Reply by jerinpetergeorge** [reply]
If I had a choice, I would go for a separate serializer for `Musician` model, as

```python
class MusicianFieldSerializer(serializers.ModelSerializer):
    class Meta:
        model = Musician
        fields = ("id", "name")


class AlbumSerializer(serializers.ModelSerializer):
    artist = MusicianFieldSerializer() # use serializer instead of field

    class Meta:
        fields = '__all__'
        model = Album
```

@tomchristie @dracarys-github

**Reply by dracarys-github** [reply]
Thanks @jerinpetergeorge for mentioning the minimum reproducible example. I have also created a git repository https://github.com/dracarys-github/sample_SO_ref clearly showing the issue

**Reply by dracarys-github** [reply]
@jerinpetergeorge I agree that Nested Serializers would work and that is a viable solution to the problem. But the fact that nested serializers work eliminates the possibility of the browsable API not having the capability or not intended to be designed for the displaying dictionaries as the value of a key in the JSON response. 

Moreover this only being a problem with the post,put or patch being used with the GenericViewAPI leads me to believe that a edge case has been missed somewhere.

**Reply by jerinpetergeorge** [reply]
I think you are doing it in the wrong way. At least, DRF doesn't expect someone to return a `dict` value from the `to_representation()` method of `RelatedField` or it's sub-classes.


The error can be get rid of by inheriting the `CustomField` class from **`serializers.Field`** rather than `serializers.RelatedField`

Ex:
```python
class CustomField(serializers.Field):
    def to_representation(self, value):
        return {"id": value.pk, "name": value.name}
```

**Reply by dracarys-github** [reply]
@jerinpetergeorge  Sorry for the late reply. 
Thanks for your solution, it works but it still is a workaround. 

I checked the DRF docs nowhere does it mention that ```to_representation()``` method of ```RelatedField``` or it's sub-classes cannot return a ```dict```. Intuitively also no attribute of ```RelatedField``` makes me think that it should not return a ```dict```.

Moreover, it perfectly works in returning the necessary JSON response whenever the API endpoint is called. It just fails to do so when the template that is required to render for the Browsable API cannot parse a ```dict```. Does that not seem like a frontend problem (template) rather than something the the ```RelatedField``` does not have the capability to handle.

@jerinpetergeorge, I might be wrong about this but it seems like a pretty common use case scenario and I feel it would really help do delve a little deeper into it rather than shunning it as a non-issue without much thought. Even if I am wrong, at the least a small note can be added in the documentation, that would help a lot of people as this is a fairly common use case. (You yourself have used it like this in some of your SO answers: https://stackoverflow.com/questions/29950956/drf-simple-foreign-key-assignment-with-nested-serializers/52246232#52246232)

**Comment by dracarys-github** [comment]
@tomchristie For your reference, here is the error image
![image](https://user-images.githubusercontent.com/83649898/118155506-ee58de80-b435-11eb-9462-715eb5e1efd2.png)
